### PR TITLE
Add the concept of "advanced" commands.

### PR DESCRIPTION
**Pull request by riannucci**
This allows a program to define many subcommands, but have them omitted from the default `help` view, to keep the high-level subcommands directly visible to the user.

### CODE REVIEW

File: subcommands.go [hunk 1]
```diff
@@ -283,23 +299,29 @@ func wrapWithLines(s string) string {
...
-	UsageLine:  "help <command>",
-	ShortDesc:  "prints help about a command",
-	LongDesc:   "Prints an overview of every commands or information about a specific command.",
-	CommandRun: func() CommandRun { return &helpRun{} },
+	UsageLine: "help <command> [-advanced]",
+	ShortDesc: "prints help about a command",
+	LongDesc:  "Prints an overview of every commands or information about a specific command.\nPass -advanced to see help for advanced commands.",
+	CommandRun: func() CommandRun {
+		ret := &helpRun{}
+		ret.advanced = ret.Flags.Bool("advanced", false, "show advanced commands")
```
Comment (vadimsht) [inline subcommands.go]: Would it be difficult to not register this flag if Application has no advanced commands defined?

File: subcommands.go [hunk 2]
```diff
@@ -118,36 +118,44 @@ func (c *Command) Name() string {
...
-
-`
-
-const usageTemplateAdvanced = `Commands:{{range .GetCommands}}
-  {{.Name | printf "%-11s"}} {{.ShortDesc}}{{end}}`
+// usage prints out the general application usage.
+//
+// TODO(maruel): Use termbox-go to enable coloring!
+func usage(out io.Writer, a Application, includeAdvanced bool) {
+	usageTemplate := `{{.Title}}
```
Comment (maruel) [inline subcommands.go:125]: It's a long constant to put inside a function, that's the reason I had used a named variable.

Comment (riannucci) [reply subcommands.go:125]: I'm not sure what you mean? It was already inside the function. The first commit in this PR moved it out, but then I realized I could make the overall diff smaller. I can squash the two changes if you like.

Comment (maruel) [reply subcommands.go:125]: Ohh I only looked at one commit by error.

File: subcommands.go [hunk 3]
```diff
@@ -283,23 +307,29 @@ func wrapWithLines(s string) string {
...
+// CmdHelp defines the help command. It should be included in your application's
+// Commands list.
 //
 // It is not added automatically but it will be run automatically if added.
 var CmdHelp = &Command{
-	UsageLine:  "help <command>",
-	ShortDesc:  "prints help about a command",
-	LongDesc:   "Prints an overview of every commands or information about a specific command.",
-	CommandRun: func() CommandRun { return &helpRun{} },
+	UsageLine: "help <command> [-advanced]",
```
Comment (maruel) [inline subcommands.go]: "help [-advanced] <command>"

File: subcommands.go [hunk 4]
```diff
@@ -283,23 +307,29 @@ func wrapWithLines(s string) string {
...
+	ShortDesc: "prints help about a command",
+	LongDesc:  "Prints an overview of every command or information about a specific command.\nPass -advanced to see help for advanced commands.",
+	CommandRun: func() CommandRun {
+		ret := &helpRun{}
+		ret.advanced = ret.Flags.Bool("advanced", false, "show advanced commands")
```
Comment (maruel) [inline subcommands.go]: ret.Flags.BoolVar(&ret.advanced, ...)  ?

File: subcommands.go [hunk 5]
```diff
@@ -199,9 +223,9 @@ func FindNearestCommand(a Application, name string) *Command {
 	}
 
 	// Calculate the levenshtein distance and take the closest one.
-	var closestD int = 1000
+	var closestD = 1000
```
Comment (maruel) [inline subcommands.go]: closestD := 1000

I'm not sure why I had written the code the way I did previously.

File: subcommands.go [hunk 6]
```diff
@@ -120,18 +121,41 @@ func (c *Command) Name() string {
 // usage prints out the general application usage.
 //
 // TODO(maruel): Use termbox-go to enable coloring!
-func usage(out io.Writer, a Application) {
-	usageTemplate := `{{.GetTitle}}
+func usage(out io.Writer, a Application, includeAdvanced bool) {
+	usageTemplate := `{{.Title}}
 
-Usage:  {{.GetName}} [command] [arguments]
+Usage:  {{.Name}} [command] [arguments]
```
Comment (maruel) [inline subcommands.go:127]: why this change?

Comment (riannucci) [reply subcommands.go:127]: Because I'm no longer passing the application object. Doesn't make sense to have the map keys named 'GetBlah'

File: subcommands.go (hunk 4 above)
Comment (riannucci) [reply subcommands.go]: _shrug_ was following the rest of the file's style

**Comment by riannucci** [comment]
Fixed, but I'll squash when it lgty.

**Comment by riannucci** [comment]
Y/N/Maybe?

**Comment by riannucci** [comment]
Squashed to make change clearer

File: subcommands.go [hunk 7]
```diff
@@ -120,18 +121,41 @@ func (c *Command) Name() string {
...
 
-Use "{{.GetName}} help [command]" for more information about a command.
+Use "{{.Name}} help [command]" for more information about a command.{{if .ShowAdvancedTip}}
+Use "{{.Name}} help -advanced" to display all commands.{{end}}
 
 `
-	tmpl(out, usageTemplate, a)
+
+	allCmds := a.GetCommands()
+	cmds := []*Command(nil)
```
Comment (maruel) [inline subcommands.go]: I prefer for uninitialize slices:
`var cmd []*Command`

(this is the only case where I prefer the var form)

**Review by maruel — APPROVED** [review]
lgtm with one style nit.

**Comment by riannucci** [comment]
Nit fixed!
